src/idom/utils.py

```diff
@@ -56,6 +62,92 @@ def __repr__(self) -> str:
         return f"{type(self).__name__}({current})"
 
 
+def vdom_to_html(value: str | VdomDict, indent: int = 0, depth: int = 0) -> str:
+    """Convert a VDOM dictionary into an HTML string
+
+    Only the following keys are translated to HTML:
+
+    - ``tagName``
+    - ``attributes``
+    - ``children`` (must be strings or more VDOM dicts)
+    """
+    if indent:
+        close_indent = f"\n{' ' * (indent * depth)}"
+        open_indent = close_indent if depth else close_indent[1:]
+    else:
+        open_indent = close_indent = ""
+
+    if isinstance(value, str):
+        return f"{open_indent}{value}" if depth else value
+
+    try:
+        tag = value["tagName"]
+    except TypeError as error:  # pragma: no cover
+        raise TypeError(f"Expected a VDOM dictionary or string, not {value}") from error
+
+    if "attributes" in value:
+        if not tag:  # pragma: no cover
+            warn("Ignored attributes from element frament", UserWarning)
+        else:
+            vdom_attributes = dict(value["attributes"])
+            if "style" in vdom_attributes:
+                vdom_attributes["style"] = _vdom_to_html_style(vdom_attributes["style"])
+            for k, v in list(vdom_attributes.items()):
+                if not isinstance(v, (str, int)):
+                    del vdom_attributes[k]
+                    warn(
+                        f"Could not convert attribute of type {type(v).__name__} to HTML attribute - {v}",
+                        UserWarning,
+                    )
+            attributes = (
+                f""" {' '.join(f'{k}="{html_escape(v)}"' for k, v in vdom_attributes.items())}"""
+                if vdom_attributes
+                else ""
+            )
+    else:
+        attributes = ""
+
+    if "children" in value:
+        children_list: list[str] = []
+
+        for child in value["children"]:
+            if isinstance(child, (dict, str)):
+                children_list.append(
+                    vdom_to_html(cast("VdomDict | str", child), indent, depth + 1)
+                )
+            else:
+                warn(
+                    f"Could not convert element of type {type(child).__name__!r} to HTML - {child}",
+                    UserWarning,
+                )
+
+        children = "".join(children_list)
+
+    else:
+        children = ""
+
+    if not children:
+        return f"{open_indent}<{tag}{attributes} />" if tag else ""
+    else:
+        return (
+            f"{open_indent}<{tag}{attributes}>{children}{close_indent}</{tag}>"
```

Comment:
The current implementation of `vdom_to_html` is currently a bit complex.

It might be cleaner and potentially less error prone to convert from

VDOM -> `lxml.etree` -> `etree.tostring`

Reply:
I think things will be simplified if I remove the indent option. If not I'll try that out.

Reply:
Much of the complication came from warning about, and then excluding, things we don't strictly know how to convert to HTML. In the latest changes, I simply cast such values to strings and HTML escape them.